src/Agent.Plugins/PipelineCache/FingerprintCreator.cs

```diff
@@ -159,189 +170,260 @@ internal static Enumeration DetermineFileEnumerationFromGlob(string includeGlobP
+                            else
                             {
-                                // Path.GetRelativePath returns 'The relative path, or path if the paths don't share the same root.'
-                                string displayPath = filePathRoot == null ? path : Path.GetRelativePath(filePathRoot, path);
-                                matchedFiles.Add(path, new MatchedFile(displayPath, fs));
+                                matchedDirectories.Add(path, displayPath);
                             }
                         }
                     }
 
-                    var patternSegment = keySegment.IndexOfAny(GlobChars) >= 0 || matchedFiles.Count() > 1;
+                    var patternSegment = segment.IndexOfAny(GlobChars) >= 0 || matchedFiles.Count > 1;
+                    hasPatternSegments |= patternSegment;
 
-                    var displayKeySegment = keySegment;
+                    var displaySegment = segment;
 
                     if (context.Container != null)
                     {
-                        displayKeySegment = context.Container.TranslateToContainerPath(displayKeySegment);
+                        displaySegment = context.Container.TranslateToContainerPath(displaySegment);
                     }
 
-                    LogKeySegment(displayKeySegment, 
-                        patternSegment ? KeySegmentType.FilePattern : KeySegmentType.FilePath,
-                        matchedFiles.Values.ToArray());
-
-                    if (!matchedFiles.Any())
+                    KeySegmentType segmentType;
+                    object details;
+                    if (fingerprintType == FingerprintType.Key)
                     {
-                        if (patternSegment)
+                        segmentType = patternSegment ? KeySegmentType.FilePattern : KeySegmentType.FilePath;
+                        details = matchedFiles.Values.ToArray();
+                        resolvedSegments.Add(MatchedFile.GenerateHash(matchedFiles.Values));
+
+                        if (!matchedFiles.Any())
                         {
-                            exceptions.Add(new FileNotFoundException($"No matching files found for pattern: {displayKeySegment}"));
+                            var message = patternSegment ? $"No matching files found for pattern: {displaySegment}" : $"File not found: {displaySegment}";
+                            exceptions.Add(new FileNotFoundException(message));
                         }
-                        else
+                    }
+                    else
+                    {
+                        segmentType = patternSegment ? KeySegmentType.DirectoryPattern : KeySegmentType.Directory;
+                        details = matchedDirectories.Values.ToArray();
+                        resolvedSegments.AddRange(matchedDirectories.Values);
+
+                        // TODO: Is it the right behavior to throw an exception if a path segment isn't resolveable?
```

Comment:
I'd love the team's input on what this behavior should be. For any given path segment, if there are no matches, should the tasks fail or simply not include it?

Reply:
hmm, We should fail imo